_posts/2023-11-13-bright-manas-partnership.md

```diff
@@ -0,0 +1,31 @@
+
+The development of these tools was sponsored by [Bright](https://brightsec.com/), maker of an intelligent exploiter for securing websites. The exploiter works by searching for endpoints and attacking them with a large set of potential security threats. As an established product, it has grown organically over the years, requiring special tools to improve it. Therefore, Bright asked Manas to help out in two major subjects: refactoring of the application and memory leak hunting.
+
+## Refactoring
+
+The application has two distinctive features: discovering endpoints, and attacking them. From an architectural point of view, it made sense to split them in two different applications. This was a major refactor that required the construction of `crystal tool dependencies`, which shows a tree of `require` dependencies. It becomes easier to split the application following the require graph.
+
+Additionally, `crystal tool dependencies` helps understanding conflicts arising from the `require` order. For instance, if two files `a.cr` and `b.cr` define the same method for a class, Crystal uses the last definition that comes the last file being required. With this tool we can observe which file is the last one required between `a.cr` and `b.cr`, and understand why.
+
+Once the separation was performed, we checked for unreachable `def`s with `crystal tool unreachable` in order to help clean up dead code.
+
+## Memory leak hunting
+
+In certain cases the application’s memory consumption could grow indefinitely until exhausting the global memory. In order to inspect memory usage and catch leaked fibers, we created two tools, available in the [perf-tools](https://github.com/crystal-lang/perf-tools) shard. The first tool is `mem_prof`, which once imported tracks memory allocations, allowing for the listing of memory allocations per location or per type. For instance, it is possible to track if there is an instance of a class that is holding a large amount of data that should be _garbage collected_ after certain point to release memory. If it is not collected, then that might lead to a leak.
+
+The second tool is `fiber_trace`, which lists the running fibers together with their allocation point and the yield point —the place in which the fiber returned execution to the scheduler. An example of use is to track if a fiber is still alive, and if so, what was the last operation that it did.
+
+Together, these two tools helped us and the Bright team fix several leaks.
+
+## Conclusions
+
+Working on an actual large application was an enriching experience, and building the necessary tools helped us improve the compiler and its ecosystem. We are very thankful to Bright for their support, and we invite companies facing difficulties with their developments to help us grow the Crystal ecosystem even further.
```

Comment:
**nitpick (stylistic):**

```suggestion
Working on an large production application was an enriching experience, and building the necessary tools helped us improve the compiler and its ecosystem. We are very thankful to Bright for their support, and we invite companies facing difficulties with their developments to help us grow the Crystal ecosystem even further.
```

Reply:
**suggestion:** I think we should highlight the last clause more. It deserves its own paragraph. And maybe a more direct address?
Rough suggestion:

```suggestion
Working on an actual large application was an enriching experience, and building the necessary tools helped us improve the compiler and its ecosystem. We are very thankful to Bright for their support.

If you are facing difficulties with your Crystal application, please reach out to crystal@manas.tech to talk about how we can help.
Maybe it ends up growing the Crystal ecosystem even further with new tools.
```

Reply:
I edited your comment ☝️  let me know what you think

Reply:
**Nitpick:** Maybe replace one `help` with support `support` to avoid duplication, but that's a tiny detail.

Reply:
There's also "support" :-)

Reply:
But not three words apart.